File: 496-next-greater-element-i/496.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        ans = []
        for num in nums1:
            i = 0
            greater = None
            while i < len(nums2):
                if nums2[i] == num:
                    i += 1
                    break
                i += 1

            while i < len(nums2):
                if nums2[i] > num:
                    greater = nums2[i]
                    break
                i += 1

            if greater:
                ans.append(greater)
            else:
                ans.append(-1)

        return ans

File: 496-next-greater-element-i/496.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        vector<int> ans;
        stack<int> s;
        map<int, int> m;

        for (int n: nums2) {
            while (!s.empty() && n > s.top()) {
                int v = s.top();
                m[v] = n;
                s.pop();
            }
            s.push(n);
        }

        while (!s.empty()) {
            int v = s.top();
            m[v] = -1;
            s.pop();
        }

        for (int n: nums1) {
            if (m.find(n) == m.end()) {
                ans.push_back(-1);
            } else {
                ans.push_back(m[n]);
            }
        }

        return ans;
    }
};
